File: tools/inspect_subtitle_delay.py
from __future__ import annotations

import argparse
from pathlib import Path

from rtve_dl.log import set_debug
from rtve_dl.subs.delay_auto import _estimate_by_asr, _estimate_by_energy
from rtve_dl.subs.srt_parse import parse_srt
from rtve_dl.subs.vtt import parse_vtt


def _format_estimate(label: str, est: object | None) -> str:
    if est is None:
        return f"{label}: none"
    return (
        f"{label}: delay_ms={est.delay_ms} confidence={est.confidence:.3f} "
        f"method={est.method} matched={est.matched}"
    )


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Inspect subtitle auto-delay estimation for one local MP4 + SRT pair."
    )
    parser.add_argument("--mp4", required=True, help="Path to source MP4")
    parser.add_argument("--srt", default=None, help="Path to Spanish SRT")
    parser.add_argument("--vtt", default=None, help="Path to Spanish RTVE VTT; use this for parity with production")
    parser.add_argument(
        "--tmp-dir",
        default="tmp/subtitle-delay-inspect",
        help="Directory for temporary ASR clip/output files",
    )
    parser.add_argument("--base", default=None, help="Base name used in temp file names")
    parser.add_argument("--max-ms", type=int, default=15000, help="Maximum absolute delay to consider")
    parser.add_argument("--skip-energy", action="store_true", help="Skip energy-based estimation")
    parser.add_argument("--skip-asr", action="store_true", help="Skip ASR-based estimation")
    parser.add_argument("--save-asr-srt", default=None, help="Optional path to save the ASR SRT used for delay estimation")
    parser.add_argument("--debug", action="store_true", help="Enable debug logging")
    parser.add_argument(
        "--asr-backend",
        default="whisperx",
        choices=["whisperx", "mlx"],
        help="Python auto-delay ASR backend",
    )
    parser.add_argument("--asr-model", default="small", help="WhisperX model name")
    parser.add_argument("--asr-device", default="cpu", help="WhisperX device")
    parser.add_argument("--asr-compute-type", default="int8", help="WhisperX compute type")
    parser.add_argument("--asr-batch-size", type=int, default=8, help="WhisperX batch size")
    parser.add_argument("--asr-vad-method", default="silero", help="WhisperX VAD method")
    parser.add_argument("--asr-mlx-model", default="mlx-community/whisper-small-mlx", help="MLX model repo")
    args = parser.parse_args()

    set_debug(args.debug)
    mp4 = Path(args.mp4)
    srt = Path(args.srt) if args.srt else None
    vtt = Path(args.vtt) if args.vtt else None
    tmp_dir = Path(args.tmp_dir)
    base = args.base or mp4.stem

    if not mp4.exists():
        raise SystemExit(f"missing mp4 file: {mp4}")
    if vtt is None and srt is None:
        raise SystemExit("one of --vtt or --srt is required")
    if vtt is not None:
        if not vtt.exists():
            raise SystemExit(f"missing vtt file: {vtt}")
        cues = parse_vtt(vtt.read_text(encoding="utf-8", errors="replace"))
        source_label = f"vtt: {vtt}"
    else:
        if srt is None or not srt.exists():
            raise SystemExit(f"missing srt file: {srt}")
        cues = parse_srt(srt.read_text(encoding="utf-8", errors="replace"))
        source_label = f"srt: {srt}"
    if not cues:
        raise SystemExit(f"no cues found in {source_label}")

    energy_est = None
    asr_est = None
    if not args.skip_energy:
        energy_est = _estimate_by_energy(cues, mp4, max_ms=max(1, args.max_ms))
    if not args.skip_asr:
        asr_est = _estimate_by_asr(
            cues=cues,
            mp4_path=mp4,
            tmp_dir=tmp_dir,
            base=base,
            asr_backend=args.asr_backend,
            asr_model=args.asr_model,
            asr_device=args.asr_device,
            asr_compute_type=args.asr_compute_type,
            asr_batch_size=args.asr_batch_size,
            asr_vad_method=args.asr_vad_method,
            asr_mlx_model=args.asr_mlx_model,
            max_ms=max(1, args.max_ms),
            save_asr_srt=Path(args.save_asr_srt) if args.save_asr_srt else None,
        )

    if energy_est is None and asr_est is None:
        final_est = None
    elif energy_est is None:
        final_est = asr_est
    elif energy_est.confidence >= 0.10 or asr_est is None:
        final_est = energy_est
    else:
        final_est = asr_est

    print(f"mp4: {mp4}")
    print(source_label)
    print(f"cues: {len(cues)}")
    print(_format_estimate("energy", energy_est))
    print(_format_estimate("asr", asr_est))
    print(_format_estimate("final", final_est))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/src/bin/inspect_subtitle_delay.rs
use anyhow::{Context, Result};
use clap::Parser;
use rtve_dl_rust::asr::AsrOptions;
use rtve_dl_rust::delay::{inspect_episode_delay, DelayEstimate};
use rtve_dl_rust::logging::set_debug;
use rtve_dl_rust::subs::{parse_srt, parse_vtt};
use std::fs;
use std::path::PathBuf;

#[derive(Parser, Debug)]
#[command(name = "inspect-subtitle-delay", about = "Inspect Rust subtitle auto-delay for one local MP4 + SRT pair")]
struct Args {
    #[arg(long)]
    mp4: PathBuf,

    #[arg(long)]
    srt: Option<PathBuf>,

    #[arg(long)]
    vtt: Option<PathBuf>,

    #[arg(long, default_value = "tmp/subtitle-delay-inspect-rust")]
    tmp_dir: PathBuf,

    #[arg(long)]
    base: Option<String>,

    #[arg(long, default_value_t = 15_000)]
    max_ms: i64,

    #[arg(long, default_value = "auto")]
    asr_backend: String,

    #[arg(long, default_value = "small")]
    asr_model: String,

    #[arg(long, default_value = "cpu")]
    asr_device: String,

    #[arg(long, default_value = "int8")]
    asr_compute_type: String,

    #[arg(long, default_value_t = 8)]
    asr_batch_size: usize,

    #[arg(long, default_value = "silero")]
    asr_vad_method: String,

    #[arg(long)]
    skip_energy: bool,

    #[arg(long)]
    skip_asr: bool,

    #[arg(long)]
    save_asr_srt: Option<PathBuf>,

    #[arg(short = 'd', long)]
    debug: bool,
}

fn format_estimate(label: &str, est: Option<DelayEstimate>) -> String {
    match est {
        Some(est) => format!(
            "{label}: delay_ms={} confidence={:.3} method={} matched={}",
            est.delay_ms, est.confidence, est.method, est.matched
        ),
        None => format!("{label}: none"),
    }
}

fn main() {
    if let Err(err) = run() {
        eprintln!("error: {err:#}");
        std::process::exit(1);
    }
}

fn run() -> Result<()> {
    let args = Args::parse();
    set_debug(args.debug);

    if !args.mp4.exists() {
        anyhow::bail!("missing mp4 file: {}", args.mp4.display());
    }
    if args.srt.is_none() && args.vtt.is_none() {
        anyhow::bail!("one of --vtt or --srt is required");
    }
    let (cues, source_label) = if let Some(vtt) = &args.vtt {
        if !vtt.exists() {
            anyhow::bail!("missing vtt file: {}", vtt.display());
        }
        (
            parse_vtt(&fs::read_to_string(vtt).context("read vtt file")?),
            format!("vtt: {}", vtt.display()),
        )
    } else {
        let srt = args.srt.as_ref().context("missing srt argument")?;
        if !srt.exists() {
            anyhow::bail!("missing srt file: {}", srt.display());
        }
        (
            parse_srt(&fs::read_to_string(srt).context("read srt file")?),
            format!("srt: {}", srt.display()),
        )
    };
    if cues.is_empty() {
        anyhow::bail!("no cues found in {source_label}");
    }

    let base = args
        .base
        .clone()
        .unwrap_or_else(|| args.mp4.file_stem().and_then(|s| s.to_str()).unwrap_or("episode").to_string());
    let tmp_dir = args.tmp_dir;
    fs::create_dir_all(&tmp_dir).context("create temp dir")?;

    let asr_opts = AsrOptions {
        backend: args.asr_backend,
        model: args.asr_model,
        device: args.asr_device,
        compute_type: args.asr_compute_type,
        batch_size: args.asr_batch_size,
        vad_method: args.asr_vad_method,
    };

    let inspection = inspect_episode_delay(
        &cues,
        &args.mp4,
        &tmp_dir,
        &base,
        args.max_ms.max(1),
        &asr_opts,
        !args.skip_energy,
        !args.skip_asr,
        args.save_asr_srt.as_deref(),
    )?;

    println!("mp4: {}", args.mp4.display());
    println!("{source_label}");
    println!("cues: {}", cues.len());
    println!("{}", format_estimate("energy", inspection.energy));
    println!("{}", format_estimate("asr", inspection.asr));
    println!("{}", format_estimate("final", inspection.final_estimate));
    Ok(())
}
